### PR TITLE
PAINTROID-291: Color changes cannot be undone or redone

### PR DESCRIPTION
**Pull request by JulianJautz**
color changes weren't considered as commands and not added to CommandManager

* implemented a ColorChangedCommand
* ColorChangedCommand is added to CommandManager when color is selected in ColorPickerDialog
* adapted undo/redo behaviour in line tool to work with ColorCommands

(https://jira.catrob.at/browse/PAINTROID-291)

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Paintroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *#paintroid* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by illo1311 — CHANGES_REQUESTED** [review]
Looks fine to me, but please resolve the merge conflict :)

**Review by illo1311 — APPROVED** [review]
lgtm

**Review by wslany — CHANGES_REQUESTED** [review]
I found several crashes, as follows: 

- On Android 8 to Android 12, when the color is changed _at least once,_ then when I save the image in the .catrobat-image format, there is an error while saving shown on the screen (this error only appears when there are color changes --- if the color stays the black one, then it works fine; also when it is a different format like png), as follows:
image

The catrobat-image file is nevertheless written to the file system, but when I try to open that image using the "Load image" option of Paintroid, the app crashes with the following stack trace:
```
05-07 18:06:03.916 E/AndroidRuntime( 1746): FATAL EXCEPTION: DefaultDispatcher-worker-3
05-07 18:06:03.916 E/AndroidRuntime( 1746): FATAL EXCEPTION: DefaultDispatcher-worker-3
05-07 18:06:03.916 E/AndroidRuntime( 1746): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 1746
05-07 18:06:03.916 E/AndroidRuntime( 1746): com.esotericsoftware.kryo.KryoException: Buffer underflow.
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.require(Input.java:219)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:407)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:145)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:758)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:869)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV3(VersionSerializer.kt:53)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV2(VersionSerializer.kt:50)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV1(VersionSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.handleVersions(VersionSerializer.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:39)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:777)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandSerializationUtilities.readFromFile(CommandSerializationUtilities.kt:194)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.getBitmapReturnValue(LoadImage.kt:65)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.access$getBitmapReturnValue(LoadImage.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage$execute$1.invokeSuspend(LoadImage.kt:99)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
05-07 18:06:03.916 E/AndroidRuntime( 1746): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 1746
05-07 18:06:03.916 E/AndroidRuntime( 1746): com.esotericsoftware.kryo.KryoException: Buffer underflow.
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.require(Input.java:219)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:407)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:145)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:758)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:869)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV3(VersionSerializer.kt:53)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV2(VersionSerializer.kt:50)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV1(VersionSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.handleVersions(VersionSerializer.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:39)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:777)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandSerializationUtilities.readFromFile(CommandSerializationUtilities.kt:194)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.getBitmapReturnValue(LoadImage.kt:65)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.access$getBitmapReturnValue(LoadImage.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage$execute$1.invokeSuspend(LoadImage.kt:99)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
05-07 18:06:03.916 E/AndroidRuntime( 1746): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 1746
05-07 18:06:03.916 E/AndroidRuntime( 1746): com.esotericsoftware.kryo.KryoException: Buffer underflow.
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.require(Input.java:219)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:407)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:145)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:758)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:869)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV3(VersionSerializer.kt:53)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV2(VersionSerializer.kt:50)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV1(VersionSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.handleVersions(VersionSerializer.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:39)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:777)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandSerializationUtilities.readFromFile(CommandSerializationUtilities.kt:194)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.getBitmapReturnValue(LoadImage.kt:65)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.access$getBitmapReturnValue(LoadImage.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage$execute$1.invokeSuspend(LoadImage.kt:99)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
05-07 18:06:03.916 E/AndroidRuntime( 1746): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 1746
05-07 18:06:03.916 E/AndroidRuntime( 1746): com.esotericsoftware.kryo.KryoException: Buffer underflow.
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.require(Input.java:219)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:407)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:145)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:758)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:869)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.readCurrentVersion(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV3(VersionSerializer.kt:53)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV2(VersionSerializer.kt:50)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.readV1(VersionSerializer.kt:47)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.VersionSerializer.handleVersions(VersionSerializer.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:39)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandManagerModelSerializer.read(CommandManagerModelSerializer.kt:27)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:777)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.command.serialization.CommandSerializationUtilities.readFromFile(CommandSerializationUtilities.kt:194)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.getBitmapReturnValue(LoadImage.kt:65)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage.access$getBitmapReturnValue(LoadImage.kt:37)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at org.catrobat.paintroid.iotasks.LoadImage$execute$1.invokeSuspend(LoadImage.kt:99)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:06:03.916 E/AndroidRuntime( 1746): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

- On Android 6, 7 and 7.1, the behavior is different: when I tap on the color button, then select a color, and tap on "Apply" in the color chooser, then the app immediately crashes with the following stack trace:
```
Load Previous
05-07 18:12:13.450 E/ClipboardServiceEx(16083): clipEx is android.sec.clipboard.ClipboardExManager@e116d5e
05-07 18:12:13.450 E/ClipboardServiceEx(16083): clipEx has text data : false
05-07 18:12:19.650 E/ViewRootImpl(16083): sendUserActionEvent() mView == null
05-07 18:12:44.800 E/AndroidRuntime(16083): FATAL EXCEPTION: DefaultDispatcher-worker-2
05-07 18:12:44.800 E/AndroidRuntime(16083): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 16083
05-07 18:12:44.800 E/AndroidRuntime(16083): android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:8134)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewRootImpl.invalidateChildInParent(ViewRootImpl.java:1257)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewGroup.invalidateChild(ViewGroup.java:5428)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidateInternal(View.java:13927)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidate(View.java:13891)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidate(View.java:13875)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.ui.tools.DefaultBrushToolOptionsView.invalidate(DefaultBrushToolOptionsView.kt:143)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.tools.implementation.BrushTool.changePaintColor(BrushTool.kt:158)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.ColorChangedCommand.run(ColorChangedCommand.kt:38)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.DefaultCommandManager.executeCommand(DefaultCommandManager.kt:85)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.DefaultCommandManager.addCommand(DefaultCommandManager.kt:77)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.AsyncCommandManager$addCommand$1.invokeSuspend(AsyncCommandManager.kt:65)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
05-07 18:12:44.800 E/ViewRootImpl(16083): sendUserActionEvent() mView == null
05-07 18:11:59.240 E/Zygote  (16083): v2
05-07 18:11:59.250 E/Zygote  (16083): accessInfo : 0
05-07 18:11:59.980 E/MotionRecognitionManager(16083): mSContextService = android.hardware.scontext.ISContextService$Stub$Proxy@bf1e03b
05-07 18:11:59.980 E/MotionRecognitionManager(16083): motionService = com.samsung.android.motion.IMotionRecognitionService$Stub$Proxy@f0a358
05-07 18:11:59.980 E/MotionRecognitionManager(16083): motionService = com.samsung.android.motion.IMotionRecognitionService$Stub$Proxy@f0a358
05-07 18:12:00.470 E/ClipboardServiceEx(16083): clipEx is android.sec.clipboard.ClipboardExManager@e116d5e
05-07 18:12:00.470 E/ClipboardServiceEx(16083): clipEx has text data : false
05-07 18:12:13.450 E/ClipboardServiceEx(16083): clipEx is android.sec.clipboard.ClipboardExManager@e116d5e
05-07 18:12:13.450 E/ClipboardServiceEx(16083): clipEx has text data : false
05-07 18:12:19.650 E/ViewRootImpl(16083): sendUserActionEvent() mView == null
05-07 18:12:44.800 E/AndroidRuntime(16083): FATAL EXCEPTION: DefaultDispatcher-worker-2
05-07 18:12:44.800 E/AndroidRuntime(16083): Process: org.catrobat.paintroid.independent_20220506_1943, PID: 16083
05-07 18:12:44.800 E/AndroidRuntime(16083): android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:8134)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewRootImpl.invalidateChildInParent(ViewRootImpl.java:1257)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.ViewGroup.invalidateChild(ViewGroup.java:5428)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidateInternal(View.java:13927)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidate(View.java:13891)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at android.view.View.invalidate(View.java:13875)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.ui.tools.DefaultBrushToolOptionsView.invalidate(DefaultBrushToolOptionsView.kt:143)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.tools.implementation.BrushTool.changePaintColor(BrushTool.kt:158)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.ColorChangedCommand.run(ColorChangedCommand.kt:38)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.DefaultCommandManager.executeCommand(DefaultCommandManager.kt:85)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.DefaultCommandManager.addCommand(DefaultCommandManager.kt:77)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at org.catrobat.paintroid.command.implementation.AsyncCommandManager$addCommand$1.invokeSuspend(AsyncCommandManager.kt:65)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
05-07 18:12:44.800 E/AndroidRuntime(16083): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
05-07 18:12:44.800 E/ViewRootImpl(16083): sendUserActionEvent() mView == null
```

File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/junit/command/DefaultCommandFactoryTest.java [hunk 1]
```diff
@@ -131,4 +141,12 @@ public void testCreateResizeCommand() {
...
 		assertThat(command, is(instanceOf(ResizeCommand.class)));
 	}
+
+	@Test
+	public void testCreateColorChangedCommand() {
+		ToolReference toolReference = new DefaultToolReference();
+		MockitoAnnotations.initMocks(this);
+		Command command = commandFactory.createColorChangedCommand(toolReference, bottomNavigationViewHolder, Color.WHITE);
+		assertThat(command, is(instanceOf(ColorChangedCommand.class)));
+	}
```
Comment (FromG) [inline Paintroid/src/androidTest/java/org/catrobat/paintroid/test/junit/command/DefaultCommandFactoryTest.java:151]: this test seems to fail. 
createColorChangedCommand wants a context as second parameter.
Therefore jenkins can't run the ticket.

**Review by FromG — CHANGES_REQUESTED** [review]

**Review by FromG — APPROVED** [review]
everything should work.

**Review by wslany — APPROVED** [review]

**Comment by wslany** [comment]
@AneiMakovec wrote:
> I took a look at PAINTROID-291 and tested the code. I used multiple different tools while changing the colors and the undo and redo operations performed as expected for both changing the color with the color picker or with the pipette in the color picker. The only thing that I noticed was that when using the pipette tool (the standalone tool, not the one in color picker) the color changes were not remembered as a command. I don't know if this is really an issue, since the ticket description mentions only the color picker and its pipette. As for the tests, I ran them locally and they passed. I also checked the test code and it seems to test multiple use cases, but I didn't find any tests regarding the pipette in the color picker. I don't know if we can even test the pipette, so I'm not sure if this is an issue either.

Thanks Anei for doing the QA check. I think you are right, it would have been better if the color-checkpoints of the pipette tool were also included. Having said that, the fix covers a large amount of the original ticket, so it now works fine in most situations. Also, regarding tests for the pipette in the color picker, it is true but at the same time it is nearly impossible to test all possible scenarios, so I guess the current test coverage is good enough. Another consideration is that @JulianJautz who implemented the fix is currently not with the team and has no time to further optimize the implementation, so based on all this I am now merging it.

Thanks everyone who was involved in fixing and giving feedback and testing this PR!

**Pull request merged**